src/lua/protobuf.lua

```diff
@@ -0,0 +1,316 @@
+    double = true,
+    float = true,
+    int32 = true,
+    int64 = true,
+    uint32 = true,
+    uint64 = true,
+    sint32 = true,
+    sint64 = true,
+    fixed32 = true,
+    fixed64 = true,
+    sfixed32 = true,
+    sfixed64 = true,
+    bool = true,
+    string = true,
+    bytes = true,
+}
+
+local protobuf_varint_types = {
+    int32 = true,
+    int64 = true,
+    uint32 = true,
+    uint64 = true,
+    sint32 = true,
+    sint64 = true,
+    bool = true,
+    enum = true,
+}
+
+local protobuf_len_types = {
+    string = true,
+    bytes = true,
+    message = true,
+}
+
+local protobuf_I32_types = {
+    fixed32 = true,
+    sfixed32 = true,
+}
+
+local protobuf_I64_types = {
+    fixed64 = true,
+    sfixed64 = true,
+}
+
+local c_integer = {
+    ['ctype<int64_t>'] = true,
+    ['ctype<uint64_t>'] = true,
+}
+
+local input_varint_types = {
+    boolean = true,
+    number = true,
+    cdata = true,
+}
+
+local input_len_types = {
+    string = true,
+}
+
+local input_I32_types = {
+    number = true,
+    cdata = true,
+}
+
+local input_I64_types = {
+    number = true,
+    cdata = true
+}
+
+-- Create a message object suitable to pass
+-- into the protobuf.protocol function.
+--
+-- Accepts a name of the message and a message
+-- definition in the following format.
+--
+-- message_def = {
+--    <field_name> = {<field_type>, <field_id>},
+--    <...>
+-- }
+local function message(message_name, message_def)
+    local field_by_name = {}
+    local field_by_id = {}
+    for field_name, def in pairs(message_def) do
+        local field_type = def[1]
+        local field_id = def[2]
+        if field_by_id[field_id] ~= nil then
+            error(('Id %d in field %q was already used'):format(field_id,
+                field_name))
+        end
+        if field_id < min_field_id or field_id > max_field_id then
+            error(('Id %d in field %q is out of range [%d; %d]'):format(
+                field_id, field_name, min_field_id, max_field_id))
+        end
+        if field_id >= reserved_field_id_min and
+           field_id <= reserved_field_id_max then
+           error(('Id %d in field %q is in reserved ' ..
+               'id range [%d, %d]'):format(field_id, field_name,
+               reserved_field_id_min, reserved_field_id_max))
+        end
+        local field_def = {
+            type = field_type,
+            name = field_name,
+            id = field_id,
+        }
+        field_by_name[field_name] = field_def
+        field_by_id[field_id] = field_def
+    end
+    return {
+        type = 'message',
+        name = message_name,
+        field_by_name = field_by_name,
+        field_by_id = field_by_id
+    }
+end
+
+-- Create a enum object suitable to pass into
+-- the protobuf.protocol function.
+--
+-- Accepts a name of an enum and an enum definition
+-- in the following format.
+--
+-- enum_def = {
+--     <value_name> = <value_id>,
+--     <...>
+-- }
+local function enum(enum_name, enum_def)
```

Comment:
Shouldn't we check that a zero id is present?

Reply:
Added a check for zero id, added corresponding test in protobuf_module_test